File: WebCat/Fetch/Browser/Bing.cs
using OpenQA.Selenium;
using WebCat.Fetch.Struct;

namespace WebCat.Fetch.Browser;

public static class Bing
{
    public static class Utils
    {
        public static Task GoToSearchPageAsync(IWebDriver driver, string query)
        {
            return driver
                .Navigate()
                .GoToUrlAsync($"https://www.bing.com/search?q={Uri.EscapeDataString(query)}");
        }

        private static SearchEngineResult? ParseSearchResult(IWebElement element)
        {
            try
            {
                var url = element.GetAttribute("href");
                var title = element.Text;
                SearchEngineResult? searchResult = new SearchEngineResult(title, url!);
                return searchResult;
            }
            catch (StaleElementReferenceException)
            {
                return null;
            }
        }

        public static SearchEngineResult[] GetSearchResults(IWebDriver driver) => driver
            .FindElements(By.CssSelector("#b_results > .b_algo h2 > a"))
            .Where(element => element.Displayed)
            .Select(ParseSearchResult)
            .Where(result => result is not null)
            .Distinct()
            .Select(result => result!.Value)
            .ToArray();
    }

    public static async Task<SearchEngineResult[]> FetchSearchResultsAsync(IWebDriver driver, string query)
    {
        await Utils.GoToSearchPageAsync(driver, query);
        return Utils.GetSearchResults(driver);
    }
}

File: WebCat/Bing.fs
module WebCat.Bing

open System
open OpenQA.Selenium

let private goToSearchPageAsync (driver: IWebDriver) (query: string) =
    driver
        .Navigate()
        .GoToUrlAsync $"https://www.bing.com/search?q={Uri.EscapeDataString query}"
    |> Async.AwaitTask

[<Struct>]
type SearchEngineResult = { Title: string; Url: string }

let private ParseSearchResult (element: IWebElement) : Option<SearchEngineResult> =
    try
        let url = element.GetAttribute "href"
        let title = element.Text
        Some { Url = url; Title = title }
    with :? StaleElementReferenceException ->
        None

let private getSearchResults (driver: IWebDriver) =
    driver.FindElements(By.CssSelector "#b_results > .b_algo h2 > a")
    |> Seq.filter _.Displayed
    |> Seq.map ParseSearchResult
    |> Seq.choose id
    |> Seq.distinct
    |> Seq.toArray

let fetchBingResultsAsync (driver: IWebDriver) (query: string)=
    async {
        do! goToSearchPageAsync driver query
        return getSearchResults driver
    }
